runtime/dart_service_isolate.h

```diff
@@ -21,10 +27,20 @@ class DartServiceIsolate {
 
   static std::string GetObservatoryUri();
 
+  // Returns a handle for the callback that can be used in
+  // RemoveServerStatusCallback
+  FML_WARN_UNUSED_RESULT static int AddServerStatusCallback(
+      ObservatoryServerStateCallback callback);
+  // Accepts the handle returned by AddServerStatusCallback
+  static void RemoveServerStatusCallback(int handle);
+
  private:
   // Native entries.
   static void NotifyServerState(Dart_NativeArguments args);
   static void Shutdown(Dart_NativeArguments args);
+
+  static int handle_count_;
+  static std::vector<std::pair<int, ObservatoryServerStateCallback>> callbacks_;
```

Comment:
I think I've covered all the accesses now.

Reply:
I'm a bit concerned that this could deadlock though.  If callback code tries to add or remove a callback (or remove itself).  Maybe a compare and swap would work instead of the mutexes, or maybe I should be looking to implement this on the dart side instead.